File: Java/GerlongElectric.java
import java.util.ArrayList; // Library untuk menggunakan ArrayList (list dinamis)
import java.util.Scanner;   // Library untuk membaca input dari keyboard

public class GerlongElectric { // Nama kelas utama

    private static ArrayList<Produk> daftarProduk = new ArrayList<>(); // List produk yang menyimpan semua data produk
    private static int nextId = 1; // Counter ID produk agar selalu unik
    private static Scanner scanner = new Scanner(System.in); // Scanner untuk input user

    // Titik masuk program
    public static void main(String[] args) {
        // Hardcode produk awal
        daftarProduk.add(new Produk(nextId++, "Laptop Gaming", "Komputer", 15000000.0));
        daftarProduk.add(new Produk(nextId++, "Smartphone Android", "Handphone", 4500000.0));
        daftarProduk.add(new Produk(nextId++, "Headphone Nirkabel", "Aksesoris", 1200000.0));
        System.out.println("Data produk awal berhasil dimuat.");

        int pilihan;

        // Perulangan menu sampai user memilih keluar
        do {
            tampilkanMenu();
            System.out.print("Masukkan pilihan Anda: ");
            
            try {
                pilihan = scanner.nextInt(); // Baca input pilihan
                scanner.nextLine(); // Buang newline biar input berikutnya aman
                
                // Jalankan fungsi sesuai pilihan user
                switch (pilihan) {
                    case 1:
                        tambahProduk(); // Create
                        break;
                    case 2:
                        tampilkanSemuaProduk(); // Read
                        break;
                    case 3:
                        updateProduk(); // Update
                        break;
                    case 4:
                        hapusProduk(); // Delete
                        break;
                    case 5:
                        cariProduk(); // Search
                        break;
                    case 6:
                        System.out.println("Terima kasih telah menggunakan aplikasi ini!");
                        break;
                    default:
                        System.out.println("Pilihan tidak valid. Silakan coba lagi.");
                        break;
                }
            } catch (java.util.InputMismatchException e) { 
                // Menangani jika input bukan angka
                System.out.println("Input tidak valid. Masukkan angka.");
                scanner.nextLine(); // Bersihkan input buffer
                pilihan = 0; // Set pilihan ke 0 biar loop lanjut
            }
        } while (pilihan != 6); // Ulangi sampai user pilih keluar
        
        scanner.close(); // Tutup scanner setelah selesai
    }

    // Menu utama
    private static void tampilkanMenu() {
        System.out.println("\n=== Menu Manajemen GerlongElectric ===");
        System.out.println("1. Tambah Produk");
        System.out.println("2. Tampilkan Semua Produk");
        System.out.println("3. Ubah Produk");
        System.out.println("4. Hapus Produk");
        System.out.println("5. Cari Produk");
        System.out.println("6. Keluar");
    }

    // Tambah produk baru
    private static void tambahProduk() {
        System.out.println("\n--- Tambah Produk Baru ---");
        System.out.print("Masukkan Nama Produk: ");
        String nama = scanner.nextLine();
        System.out.print("Masukkan Jenis Produk: ");
        String jenis = scanner.nextLine();
        System.out.print("Masukkan Harga: ");
        
        try {
            double harga = scanner.nextDouble();
            scanner.nextLine();
            
            if (harga < 0) { // Validasi harga
                System.out.println("Harga tidak boleh negatif. Produk tidak ditambahkan.");
            } else {
                Produk newProduk = new Produk(nextId++, nama, jenis, harga);
                daftarProduk.add(newProduk);
                System.out.println("Produk berhasil ditambahkan dengan ID " + newProduk.getId() + ".");
            }
        } catch (java.util.InputMismatchException e) {
            // Menangani input harga yang bukan angka
            System.out.println("Input harga tidak valid. Produk tidak ditambahkan.");
            scanner.nextLine();
        }
    }

    // Tampilkan semua produk
    private static void tampilkanSemuaProduk() {
        if (daftarProduk.isEmpty()) {
            System.out.println("\nTidak ada produk yang tersedia.");
        } else {
            System.out.println("\n=== Daftar Semua Produk ===");
            for (Produk produk : daftarProduk) {
                produk.display(); // Panggil metode display milik Produk
            }
        }
    }

    // Update produk berdasarkan ID
    private static void updateProduk() {
        System.out.println("\n--- Ubah Produk ---");
        System.out.print("Masukkan ID Produk yang ingin diubah: ");
        
        try {
            int idToUpdate = scanner.nextInt();
            scanner.nextLine();
            
            Produk produkToUpdate = cariProdukById(idToUpdate); // Cari produk dulu
            if (produkToUpdate != null) {
                // Input data baru
                System.out.print("Masukkan Nama baru: ");
                String newNama = scanner.nextLine();
                System.out.print("Masukkan Jenis baru: ");
                String newJenis = scanner.nextLine();
                System.out.print("Masukkan Harga baru: ");
                double newHarga = scanner.nextDouble();
                scanner.nextLine();

                // Update data produk
                produkToUpdate.setNama(newNama);
                produkToUpdate.setJenis(newJenis);
                if (produkToUpdate.setHarga(newHarga)) {
                    System.out.println("Produk dengan ID " + idToUpdate + " berhasil diubah.");
                } else {
                    System.out.println("Gagal: Harga tidak boleh negatif. Perubahan dibatalkan.");
                }
            } else {
                System.out.println("Produk dengan ID " + idToUpdate + " tidak ditemukan.");
            }
        } catch (java.util.InputMismatchException e) {
            System.out.println("Input ID tidak valid. Silakan masukkan angka.");
            scanner.nextLine();
        }
    }

    // Hapus produk berdasarkan ID
    private static void hapusProduk() {
        System.out.println("\n--- Hapus Produk ---");
        System.out.print("Masukkan ID Produk yang ingin dihapus: ");
        
        try {
            int idToDelete = scanner.nextInt();
            scanner.nextLine();
            
            Produk produkToDelete = cariProdukById(idToDelete);
            if (produkToDelete != null) {
                daftarProduk.remove(produkToDelete);
                System.out.println("Produk dengan ID " + idToDelete + " berhasil dihapus.");
            } else {
                System.out.println("Produk dengan ID " + idToDelete + " tidak ditemukan.");
            }
        } catch (java.util.InputMismatchException e) {
            System.out.println("Input ID tidak valid. Silakan masukkan angka.");
            scanner.nextLine();
        }
    }
    
    // Cari produk berdasarkan ID
    private static void cariProduk() {
        System.out.println("\n--- Cari Produk ---");
        System.out.print("Masukkan ID Produk yang ingin dicari: ");
        
        try {
            int idToFind = scanner.nextInt();
            scanner.nextLine();
            
            Produk foundProduk = cariProdukById(idToFind);
            if (foundProduk != null) {
                System.out.println("\nProduk ditemukan:");
                foundProduk.display();
            } else {
                System.out.println("Produk dengan ID " + idToFind + " tidak ditemukan.");
            }
        } catch (java.util.InputMismatchException e) {
            System.out.println("Input ID tidak valid. Silakan masukkan angka.");
            scanner.nextLine();
        }
    }
    
    // Fungsi pembantu untuk mencari produk di list berdasarkan ID
    private static Produk cariProdukById(int id) {
        for (Produk produk : daftarProduk) {
            if (produk.getId() == id) {
                return produk; // Kalau ketemu, return objek produk
            }
        }
        return null; // Kalau tidak ketemu, return null
    }
}

File: Python/GerlongElectric.py
import os

# ===============================
# Kelas Produk
# Merepresentasikan produk elektronik
# ===============================
class Produk:
    """
    Contoh penerapan OOP di Python:
    - Enkapsulasi dengan atribut private (konvensi: underscore)
    - Getter & Setter menggunakan property
    """

    def __init__(self, id, nama, jenis, harga):
        # Konstruktor: inisialisasi atribut produk
        self._id = id
        self._nama = nama
        self._jenis = jenis
        self._harga = harga

    # Getter hanya-baca untuk id (tidak ada setter agar unik)
    @property
    def id(self):
        return self._id

    # Getter & Setter untuk nama
    @property
    def nama(self):
        return self._nama

    @nama.setter
    def nama(self, new_nama):
        self._nama = new_nama

    # Getter & Setter untuk jenis
    @property
    def jenis(self):
        return self._jenis

    @jenis.setter
    def jenis(self, new_jenis):
        self._jenis = new_jenis

    # Getter & Setter untuk harga dengan validasi (>=0)
    @property
    def harga(self):
        return self._harga

    @harga.setter
    def harga(self, new_harga):
        if new_harga >= 0:
            self._harga = new_harga
        else:
            # Kalau harga negatif → raise exception
            raise ValueError("Harga tidak boleh negatif.")

    # __str__ dipanggil saat objek dicetak (print)
    def __str__(self):
        return f"""
------------------------
ID: {self._id}
Nama: {self._nama}
Jenis: {self._jenis}
Harga: {self._harga}
------------------------
"""

# ===============================
# Variabel global untuk CRUD
# ===============================
daftar_produk = []  # list penyimpanan produk
next_id = 1         # ID auto increment

# ===============================
# Fungsi CRUD dan menu
# ===============================
def tampilkan_menu():
    """Tampilkan menu utama."""
    print("\n=== Menu Manajemen GerlongElectric ===")
    print("1. Tambah Produk")
    print("2. Tampilkan Semua Produk")
    print("3. Ubah Produk")
    print("4. Hapus Produk")
    print("5. Cari Produk")
    print("6. Keluar")

# CREATE
def tambah_produk():
    global next_id
    print("\n--- Tambah Produk Baru ---")
    nama = input("Masukkan Nama Produk: ")
    jenis = input("Masukkan Jenis Produk: ")
    try:
        harga = float(input("Masukkan Harga: "))
        if harga < 0:
            print("Harga tidak boleh negatif. Produk tidak ditambahkan.")
            return
        
        # Buat objek baru lalu masukkan ke daftar
        new_produk = Produk(next_id, nama, jenis, harga)
        daftar_produk.append(new_produk)
        next_id += 1
        print(f"Produk berhasil ditambahkan dengan ID {new_produk.id}.")
    except ValueError:
        print("Input harga tidak valid. Produk tidak ditambahkan.")

# READ
def tampilkan_semua_produk():
    if not daftar_produk:
        print("\nTidak ada produk yang tersedia.")
        return
    print("\n=== Daftar Semua Produk ===")
    for produk in daftar_produk:
        print(produk)

# UPDATE
def update_produk():
    print("\n--- Ubah Produk ---")
    try:
        id_to_update = int(input("Masukkan ID Produk yang ingin diubah: "))
        produk_ditemukan = False
        for produk in daftar_produk:
            if produk.id == id_to_update:
                print("Produk ditemukan. Masukkan data baru:")
                produk.nama = input("Masukkan Nama baru: ")
                produk.jenis = input("Masukkan Jenis baru: ")
                try:
                    new_harga = float(input("Masukkan Harga baru: "))
                    produk.harga = new_harga  # setter validasi harga
                    print(f"Produk dengan ID {id_to_update} berhasil diubah.")
                except ValueError as e:
                    print(f"Gagal: {e}. Perubahan dibatalkan.")
                produk_ditemukan = True
                break
        if not produk_ditemukan:
            print(f"Produk dengan ID {id_to_update} tidak ditemukan.")
    except ValueError:
        print("Input ID tidak valid. Silakan masukkan angka.")

# DELETE
def hapus_produk():
    print("\n--- Hapus Produk ---")
    try:
        id_to_delete = int(input("Masukkan ID Produk yang ingin dihapus: "))
        produk_dihapus = False
        for produk in daftar_produk:
            if produk.id == id_to_delete:
                daftar_produk.remove(produk)
                print(f"Produk dengan ID {id_to_delete} berhasil dihapus.")
                produk_dihapus = True
                break
        if not produk_dihapus:
            print(f"Produk dengan ID {id_to_delete} tidak ditemukan.")
    except ValueError:
        print("Input ID tidak valid. Silakan masukkan angka.")

# SEARCH
def cari_produk():
    print("\n--- Cari Produk ---")
    try:
        id_to_find = int(input("Masukkan ID Produk yang ingin dicari: "))
        produk_ditemukan = False
        for produk in daftar_produk:
            if produk.id == id_to_find:
                print("\nProduk ditemukan:")
                print(produk)
                produk_ditemukan = True
                break
        if not produk_ditemukan:
            print(f"Produk dengan ID {id_to_find} tidak ditemukan.")
    except ValueError:
        print("Input ID tidak valid. Silakan masukkan angka.")

# ===============================
# Main loop
# ===============================
def main():
    global next_id
    # Tambahkan produk default
    daftar_produk.append(Produk(next_id, "Laptop Gaming", "Komputer", 15000000.0)); next_id += 1
    daftar_produk.append(Produk(next_id, "Smartphone Android", "Handphone", 4500000.0)); next_id += 1
    daftar_produk.append(Produk(next_id, "Headphone Nirkabel", "Aksesoris", 1200000.0)); next_id += 1
    print("Data produk awal berhasil dimuat.")

    # Menu loop
    while True:
        tampilkan_menu()
        pilihan = input("Masukkan pilihan Anda: ")
        if pilihan == '1': tambah_produk()
        elif pilihan == '2': tampilkan_semua_produk()
        elif pilihan == '3': update_produk()
        elif pilihan == '4': hapus_produk()
        elif pilihan == '5': cari_produk()
        elif pilihan == '6':
            print("Terima kasih telah menggunakan aplikasi ini!")
            break
        else:
            print("Pilihan tidak valid. Silakan coba lagi.")

if __name__ == "__main__":
    main()
